#データエンジニアリング入門
##第1回


###講師
B2 好田
###TA
B3 勝又


#2. SQLとは？
SQLとは、「Structured Query Language」の略で、直訳すると「構造化問い合わせ言語」という意味です。リレーショナルデータベース(RDB)（表形式のテーブル）のデータを操作するための言語です。<br>
(【SQL入門】データベース言語の基礎知識を学んでMySQL環境で実践しよう！, https://udemy.benesse.co.jp/development/system/intro-sql.html より)

# 3. 実際にやってみよう！


In [ ]:
#Setup
!pip install jupysql duckdb-engine
!wget https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv
!wget https://www.sqlitetutorial.net/wp-content/uploads/2018/03/chinook.zip
!unzip chinook.zip
%load_ext sql
%sql duckdb://

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.6/87.6 kB 2.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.1/43.1 kB 3.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 312.1/312.1 kB 9.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 376.5/376.5 kB 29.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 53.1 MB/s eta 0:00:00
--2023-09-27 07:11:01--  https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13478 (13K) [text/plain]
Saving to: ‘penguins.csv’

penguins.csv        100%[===================>]  13.16K  --.-KB/s    in 0.001s  

2023-09-27 07:11:01 (20.8 MB/s) - ‘penguins.csv’ saved [13478/13478]

--2023-09-27 07:

Connecting to 'duckdb://'

例にpenguins.csvというデータを使ってみます。

| species  |   island    | bill_length_mm |bill_depth_mm|flipper_length_mm|body_mass_g|sex|
|:---:|:---------:|:--:|:--:|:--:|:--:|:--:|
|Adelie|Torgersen|39.1|18.7|181|3750|MALE|
|Adelie|Torgersen|39.5|17.4|186|3800|FEMALE|
|Adelie|Torgersen|40.3|18|195|3250|FEMALE|
|Adelie|Torgersen||||||
|Adelie|Torgersen|36.7|19.3|193|3450|FEMALE|
|Adelie|Torgersen|39.3|20.6|190|3650|MALE|
|Adelie|Torgersen|38.9|17.8|181|3625|FEMALE|
|Adelie|Torgersen|39.2|19.6|195|4675|MALE|
|Adelie|Torgersen|34.1|18.1|193|3475||
|Adelie|Torgersen|42|20.2|190|4250||
|Adelie|Torgersen|37.8|17.1|186|3300||
|Adelie|Torgersen|37.8|17.3|180|3700||
|Adelie|Torgersen|41.1|17.6|182|3200|FEMALE|
|Adelie|Torgersen|38.6|21.2|191|3800|MALE|
|Adelie|Torgersen|34.6|21.1|198|4400|MALE|
|Adelie|Torgersen|36.6|17.8|185|3700|FEMALE|
|Adelie|Torgersen|38.7|19|195|3450|FEMALE|
|Adelie|Torgersen|42.5|20.7|197|4500|MALE|
|Adelie|Torgersen|34.4|18.4|184|3325|FEMALE|
etc...

## 3.1 SELECT / FROM
- <b>SELECT</b>
  - 参照する列を指定する
  - 列名をカンマ区切りで列挙する
  - ```*```で全ての列名を選択することができる
- <b>FROM</b>
  - 参照するテーブルを指定する
- <b>コメントアウト</b>
  - 一行のコメントアウトは頭に```-- コメント```を付ける
  - 複数行のコメントアウトは```/* コメント */```で囲う

```
SELECT
  (カラム1),（カラム2）,（カラム3）
FROM
  (テーブル名)
```

###例：


一つのコラムを抽出したいとき

In [ ]:
%%sql
SELECT
 species
FROM
 penguins.csv

Running query in 'duckdb://'

+---------+
| species |
+---------+
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
|  Adelie |
+---------+
Truncated to displaylimit of 10.

複数のコラムを抽出したいとき

In [ ]:
%%sql
SELECT
species,island,bill_length_mm
FROM
penguins.csv

Running query in 'duckdb://'

+---------+-----------+----------------+
| species |   island  | bill_length_mm |
+---------+-----------+----------------+
|  Adelie | Torgersen |      39.1      |
|  Adelie | Torgersen |      39.5      |
|  Adelie | Torgersen |      40.3      |
|  Adelie | Torgersen |      None      |
|  Adelie | Torgersen |      36.7      |
|  Adelie | Torgersen |      39.3      |
|  Adelie | Torgersen |      38.9      |
|  Adelie | Torgersen |      39.2      |
|  Adelie | Torgersen |      34.1      |
|  Adelie | Torgersen |      42.0      |
+---------+-----------+----------------+
Truncated to displaylimit of 10.

全てのコラムを抽出したいとき

In [ ]:
%%sql
SELECT
 *
FROM
 penguins.csv

Running query in 'duckdb://'

+---------+-----------+----------------+---------------+-------------------+-------------+--------+
| species |   island  | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
|  Adelie | Torgersen |      39.1      |      18.7     |        181        |     3750    |  MALE  |
|  Adelie | Torgersen |      39.5      |      17.4     |        186        |     3800    | FEMALE |
|  Adelie | Torgersen |      40.3      |      18.0     |        195        |     3250    | FEMALE |
|  Adelie | Torgersen |      None      |      None     |        None       |     None    |  None  |
|  Adelie | Torgersen |      36.7      |      19.3     |        193        |     3450    | FEMALE |
|  Adelie | Torgersen |      39.3      |      20.6     |        190        |     3650    |  MALE  |
|  Adelie | Torgersen |      38.9      |      17.8     |        181        |     3625    | FEMALE |
|  Adelie | Torgersen |      39.2      |      19.6     |        195        |     4675    |  MALE  |
|  Adelie | Torgersen |      34.1      |      18.1     |        193        |     3475    |  None  |
|  Adelie | Torgersen |      42.0      |      20.2     |        190        |     4250    |  None  |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

##3.2 演習問題
問１ penguins.csvからspecies, island, flipper_length_mm, sexのカラムを抽出してください

In [ ]:
%%sql


Running query in 'duckdb://'

+---------+-----------+-------------------+--------+
| species |   island  | flipper_length_mm |  sex   |
+---------+-----------+-------------------+--------+
|  Adelie | Torgersen |        181        |  MALE  |
|  Adelie | Torgersen |        186        | FEMALE |
|  Adelie | Torgersen |        195        | FEMALE |
|  Adelie | Torgersen |        None       |  None  |
|  Adelie | Torgersen |        193        | FEMALE |
|  Adelie | Torgersen |        190        |  MALE  |
|  Adelie | Torgersen |        181        | FEMALE |
|  Adelie | Torgersen |        195        |  MALE  |
|  Adelie | Torgersen |        193        |  None  |
|  Adelie | Torgersen |        190        |  None  |
+---------+-----------+-------------------+--------+
Truncated to displaylimit of 10.

問2 username.csvからすべてのカラムを表示してください

In [ ]:
%%sql

Running query in 'duckdb://'

+-----------+------------+------------+-----------+
|  Username | Identifier | First name | Last name |
+-----------+------------+------------+-----------+
|  booker12 |    9012    |   Rachel   |   Booker  |
|   grey07  |    2070    |   Laura    |    Grey   |
| johnson81 |    4081    |   Craig    |  Johnson  |
| jenkins46 |    9346    |    Mary    |  Jenkins  |
|  smith79  |    5079    |   Jamie    |   Smith   |
+-----------+------------+------------+-----------+

問3 username_password_recovery_code.csvからUsername, Identifier, One_time_password, Recovery_codeを表示してください




In [ ]:
%%sql

Running query in 'duckdb://'

+-----------+------------+-------------------+---------------+
|  Username | Identifier | One_time_password | Recovery_code |
+-----------+------------+-------------------+---------------+
|  booker12 |    9012    |       12se74      |     rb9012    |
|   grey07  |    2070    |       04ap67      |     lg2070    |
| johnson81 |    4081    |       30no86      |     cj4081    |
| jenkins46 |    9346    |       14ju73      |     mj9346    |
|  smith79  |    5079    |       09ja61      |     js5079    |
+-----------+------------+-------------------+---------------+

問4 username-password-recovery-code.csvからLogin email, First name, Last nameを表示してください

In [ ]:
%%sql

Running query in 'duckdb://'

+--------------------+------------+-----------+
|    Login_email     | First_name | Last_name |
+--------------------+------------+-----------+
| rachel@example.com |   Rachel   |   Booker  |
| laura@example.com  |   Laura    |    Grey   |
| craig@example.com  |   Craig    |  Johnson  |
|  mary@example.com  |    Mary    |  Jenkins  |
| jamie@example.com  |   Jamie    |   Smith   |
+--------------------+------------+-----------+

## 3.3 WHERE
- <b>WHERE</b>
  - あるカラムに対して、条件を指定することができる
  - 複数の条件を```AND```や```OR```で繋げることができる

<b>WHEREで使える演算子</b>

|記号|意味|例|
|---|---|---|
|=、!=|等しい、等しくない|```WHERE speices = "Adelie"```|
|<、>、<=、>=|大小比較|```WHERE bill_length_mm >= 3```|
|AND、OR|複数の条件を結合|```WHERE Sex = "MALE" AND species = "Adelie"```|
|IS NULL、IS NOT NULL|値がNULLかどうかの判定|```WHERE bill_length_mm IS NOT NULL```|

※その他、BETWEEN、IN、LIKEなどもあります

```
SELECT
  (カラム名)
FROM
  (テーブル名)
WHERE
  (条件式)
  ```
  

###例：


In [ ]:
%%sql
SELECT
 *
FROM
 penguins.csv
WHERE
 species='Gentoo'

Running query in 'duckdb://'

+---------+--------+----------------+---------------+-------------------+-------------+--------+
| species | island | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+--------+----------------+---------------+-------------------+-------------+--------+
|  Gentoo | Biscoe |      46.1      |      13.2     |        211        |     4500    | FEMALE |
|  Gentoo | Biscoe |      50.0      |      16.3     |        230        |     5700    |  MALE  |
|  Gentoo | Biscoe |      48.7      |      14.1     |        210        |     4450    | FEMALE |
|  Gentoo | Biscoe |      50.0      |      15.2     |        218        |     5700    |  MALE  |
|  Gentoo | Biscoe |      47.6      |      14.5     |        215        |     5400    |  MALE  |
|  Gentoo | Biscoe |      46.5      |      13.5     |        210        |     4550    | FEMALE |
|  Gentoo | Biscoe |      45.4      |      14.6     |        211        |     4800    | FEMALE |
|  Gentoo | Biscoe |      46.7      |      15.3     |        219        |     5200    |  MALE  |
|  Gentoo | Biscoe |      43.3      |      13.4     |        209        |     4400    | FEMALE |
|  Gentoo | Biscoe |      46.8      |      15.4     |        215        |     5150    |  MALE  |
+---------+--------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

In [ ]:
%%sql
SELECT
 *
FROM
 penguins.csv
WHERE
 species='Adelie' AND island='Dream'


Running query in 'duckdb://'

+---------+--------+----------------+---------------+-------------------+-------------+--------+
| species | island | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+--------+----------------+---------------+-------------------+-------------+--------+
|  Adelie | Dream  |      39.5      |      16.7     |        178        |     3250    | FEMALE |
|  Adelie | Dream  |      37.2      |      18.1     |        178        |     3900    |  MALE  |
|  Adelie | Dream  |      39.5      |      17.8     |        188        |     3300    | FEMALE |
|  Adelie | Dream  |      40.9      |      18.9     |        184        |     3900    |  MALE  |
|  Adelie | Dream  |      36.4      |      17.0     |        195        |     3325    | FEMALE |
|  Adelie | Dream  |      39.2      |      21.1     |        196        |     4150    |  MALE  |
|  Adelie | Dream  |      38.8      |      20.0     |        190        |     3950    |  MALE  |
|  Adelie | Dream  |      42.2      |      18.5     |        180        |     3550    | FEMALE |
|  Adelie | Dream  |      37.6      |      19.3     |        181        |     3300    | FEMALE |
|  Adelie | Dream  |      39.8      |      19.1     |        184        |     4650    |  MALE  |
+---------+--------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

In [ ]:
%%sql
SELECT
 *
FROM
 penguins.csv
WHERE
 island='Dream' OR body_mass_g<3900


Running query in 'duckdb://'

+---------+-----------+----------------+---------------+-------------------+-------------+--------+
| species |   island  | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
|  Adelie | Torgersen |      39.1      |      18.7     |        181        |     3750    |  MALE  |
|  Adelie | Torgersen |      39.5      |      17.4     |        186        |     3800    | FEMALE |
|  Adelie | Torgersen |      40.3      |      18.0     |        195        |     3250    | FEMALE |
|  Adelie | Torgersen |      36.7      |      19.3     |        193        |     3450    | FEMALE |
|  Adelie | Torgersen |      39.3      |      20.6     |        190        |     3650    |  MALE  |
|  Adelie | Torgersen |      38.9      |      17.8     |        181        |     3625    | FEMALE |
|  Adelie | Torgersen |      34.1      |      18.1     |        193        |     3475    |  None  |
|  Adelie | Torgersen |      37.8      |      17.1     |        186        |     3300    |  None  |
|  Adelie | Torgersen |      37.8      |      17.3     |        180        |     3700    |  None  |
|  Adelie | Torgersen |      41.1      |      17.6     |        182        |     3200    | FEMALE |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

##3.4 演習問題
問１ penguins.csvからspeciesがChinstrapまたはGentooなものを表示してください

In [ ]:
%%sql

Running query in 'duckdb://'

+-----------+--------+----------------+---------------+-------------------+-------------+--------+
|  species  | island | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+-----------+--------+----------------+---------------+-------------------+-------------+--------+
| Chinstrap | Dream  |      46.5      |      17.9     |        192        |     3500    | FEMALE |
| Chinstrap | Dream  |      50.0      |      19.5     |        196        |     3900    |  MALE  |
| Chinstrap | Dream  |      51.3      |      19.2     |        193        |     3650    |  MALE  |
| Chinstrap | Dream  |      45.4      |      18.7     |        188        |     3525    | FEMALE |
| Chinstrap | Dream  |      52.7      |      19.8     |        197        |     3725    |  MALE  |
| Chinstrap | Dream  |      45.2      |      17.8     |        198        |     3950    | FEMALE |
| Chinstrap | Dream  |      46.1      |      18.2     |        178        |     3250    | FEMALE |
| Chinstrap | Dream  |      51.3      |      18.2     |        197        |     3750    |  MALE  |
| Chinstrap | Dream  |      46.0      |      18.9     |        195        |     4150    | FEMALE |
| Chinstrap | Dream  |      51.3      |      19.9     |        198        |     3700    |  MALE  |
+-----------+--------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

問2 penguins.csvからsexが記入されているものを表示してください

In [ ]:
%%sql

Running query in 'duckdb://'

+---------+-----------+----------------+---------------+-------------------+-------------+--------+
| species |   island  | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
|  Adelie | Torgersen |      39.1      |      18.7     |        181        |     3750    |  MALE  |
|  Adelie | Torgersen |      39.5      |      17.4     |        186        |     3800    | FEMALE |
|  Adelie | Torgersen |      40.3      |      18.0     |        195        |     3250    | FEMALE |
|  Adelie | Torgersen |      36.7      |      19.3     |        193        |     3450    | FEMALE |
|  Adelie | Torgersen |      39.3      |      20.6     |        190        |     3650    |  MALE  |
|  Adelie | Torgersen |      38.9      |      17.8     |        181        |     3625    | FEMALE |
|  Adelie | Torgersen |      39.2      |      19.6     |        195        |     4675    |  MALE  |
|  Adelie | Torgersen |      41.1      |      17.6     |        182        |     3200    | FEMALE |
|  Adelie | Torgersen |      38.6      |      21.2     |        191        |     3800    |  MALE  |
|  Adelie | Torgersen |      34.6      |      21.1     |        198        |     4400    |  MALE  |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

問3 penguins.csvからsexが記入されているものであり、flipper length mmが180より小さいものを表示してください




In [ ]:
%%sql

Running query in 'duckdb://'

+---------+-----------+----------------+---------------+-------------------+-------------+--------+
| species |   island  | bill_length_mm | bill_depth_mm | flipper_length_mm | body_mass_g |  sex   |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
|  Adelie | Torgersen |      39.1      |      18.7     |        181        |     3750    |  MALE  |
|  Adelie | Torgersen |      39.5      |      17.4     |        186        |     3800    | FEMALE |
|  Adelie | Torgersen |      38.9      |      17.8     |        181        |     3625    | FEMALE |
|  Adelie | Torgersen |      41.1      |      17.6     |        182        |     3200    | FEMALE |
|  Adelie | Torgersen |      36.6      |      17.8     |        185        |     3700    | FEMALE |
|  Adelie | Torgersen |      34.4      |      18.4     |        184        |     3325    | FEMALE |
|  Adelie |   Biscoe  |      37.8      |      18.3     |        174        |     3400    | FEMALE |
|  Adelie |   Biscoe  |      37.7      |      18.7     |        180        |     3600    |  MALE  |
|  Adelie |   Biscoe  |      35.9      |      19.2     |        189        |     3800    | FEMALE |
|  Adelie |   Biscoe  |      38.2      |      18.1     |        185        |     3950    |  MALE  |
+---------+-----------+----------------+---------------+-------------------+-------------+--------+
Truncated to displaylimit of 10.

問4 penguins.csvからsex,body_mass_gが記入されているものであり、body_mass_gが3800より小さいものでspecies, body_mass_g,sexを表示してください

In [ ]:
%%sql

Running query in 'duckdb://'

+---------+-------------+--------+
| species | body_mass_g |  sex   |
+---------+-------------+--------+
|  Adelie |     3750    |  MALE  |
|  Adelie |     3250    | FEMALE |
|  Adelie |     3450    | FEMALE |
|  Adelie |     3650    |  MALE  |
|  Adelie |     3625    | FEMALE |
|  Adelie |     3200    | FEMALE |
|  Adelie |     3700    | FEMALE |
|  Adelie |     3450    | FEMALE |
|  Adelie |     3325    | FEMALE |
|  Adelie |     3400    | FEMALE |
+---------+-------------+--------+
Truncated to displaylimit of 10.